tests/unit/misc/test_mhtml.py

```diff
@@ -0,0 +1,268 @@
+    """
+
+    def __init__(self):
+        self.fp = io.BytesIO()
+
+    @property
+    def value(self):
+        return self.fp.getvalue()
+
+    def expect(self, expected):
+        actual = self.value.decode("ascii")
+        # Make sure there are no stray \r or \n
+        assert re.search(r"\r[^\n]", actual) is None
+        assert re.search(r"[^\r]\n", actual) is None
+        actual = actual.replace("\r\n", "\n")
+        expected = textwrap.dedent(expected).lstrip("\n")
+        assert expected == actual
+
+
+@pytest.fixture
+def checker():
+    return Checker()
+
+
+def test_quoted_printable_umlauts(checker):
+    content = "Die süße Hündin läuft in die Höhle des Bären"
+    content = content.encode("iso-8859-1")
+    writer = mhtml.MHTMLWriter(root_content=content,
+                               content_location="localhost",
+                               content_type="text/plain")
+    writer.write_to(checker.fp)
+    checker.expect("""
+    Content-Type: multipart/related; boundary="---=_qute-UUID"
+    MIME-Version: 1.0
+
+    -----=_qute-UUID
+    Content-Location: localhost
+    MIME-Version: 1.0
+    Content-Type: text/plain
+    Content-Transfer-Encoding: quoted-printable
+
+    Die=20s=FC=DFe=20H=FCndin=20l=E4uft=20in=20die=20H=F6hle=20des=20B=E4ren
+    -----=_qute-UUID--
+    """)
+
+
+@pytest.mark.parametrize("header, value", [
+    ("content_location", "http://brötli.com"),
+    ("content_type", "text/pläin"),
+])
+def test_refuses_non_ascii_header_value(checker, header, value):
+    defaults = {
+        "root_content": b"",
+        "content_location": "http://example.com",
+        "content_type": "text/plain",
+    }
+    defaults[header] = value
+    writer = mhtml.MHTMLWriter(**defaults)
+    with pytest.raises(UnicodeEncodeError) as e:
+        writer.write_to(checker.fp)
+        assert "'ascii' codec can't encode" in str(e)
+
+
+def test_file_encoded_as_base64(checker):
+    content = b"Image file attached"
+    writer = mhtml.MHTMLWriter(root_content=content, content_type="text/plain",
+                               content_location="http://example.com")
+    writer.add_file(location="http://a.example.com/image.png",
+                    content="\U0001F601 image data".encode("utf-8"),
+                    content_type="image/png",
+                    transfer_encoding=mhtml.E_BASE64)
+    writer.write_to(checker.fp)
+    checker.expect("""
+    Content-Type: multipart/related; boundary="---=_qute-UUID"
+    MIME-Version: 1.0
+
+    -----=_qute-UUID
+    Content-Location: http://example.com
+    MIME-Version: 1.0
+    Content-Type: text/plain
+    Content-Transfer-Encoding: quoted-printable
+
+    Image=20file=20attached
+    -----=_qute-UUID
+    Content-Location: http://a.example.com/image.png
+    MIME-Version: 1.0
+    Content-Type: image/png
+    Content-Transfer-Encoding: base64
+
+    8J+YgSBpbWFnZSBkYXRh
+
+    -----=_qute-UUID--
+    """)
```

Comment:
Minor thing, but please indent those multi-line strings one more - I find it quite hard to read them this way.